src/libstd/sys/unix/ext/net.rs

```diff
@@ -245,6 +304,340 @@ impl fmt::Debug for SocketAddr {
+    ///
+    /// ```no_run
+    /// # #![allow(unused_mut)]
+    /// #![feature(unix_socket_ancillary_data)]
+    /// use std::os::unix::net::SocketAncillary;
+    /// let mut ancillary_buffer = [0; 128];
+    /// let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
+    /// ```
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        unsafe {
+            libc::memset(buffer.as_mut_ptr().cast(), 0, buffer.len());
+        }
+        SocketAncillary { buffer, length: 0, truncated: false }
+    }
+
+    /// Returns the capacity of the buffer.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn capacity(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns the number of used bytes.
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    /// Is `true` if during a recv operation the ancillary was truncated.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// #![feature(unix_socket_ancillary_data)]
+    /// use std::os::unix::net::{UnixStream, SocketAncillary};
+    /// use std::io::IoSliceMut;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let sock = UnixStream::connect("/tmp/sock")?;
+    ///
+    ///     let mut ancillary_buffer = [0; 128];
+    ///     let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
+    ///
+    ///     let mut buf = [1; 8];
+    ///     let mut bufs = &mut [IoSliceMut::new(&mut buf[..])][..];
+    ///     sock.recv_vectored_with_ancillary(bufs, &mut ancillary)?;
+    ///
+    ///     println!("Is truncated: {}", ancillary.truncated());
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn truncated(&self) -> bool {
+        self.truncated
+    }
+
+    /// Get some file descriptors from this ancillary data into the specified buffer,
+    /// returning how many file descriptors were set.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// #![feature(unix_socket_ancillary_data)]
+    /// use std::os::unix::net::{UnixStream, SocketAncillary};
+    /// use std::io::IoSliceMut;
+    ///
+    /// fn main() -> std::io::Result<()> {
+    ///     let sock = UnixStream::connect("/tmp/sock")?;
+    ///
+    ///     let mut fds = [0; 8];
+    ///     let mut ancillary_buffer = [0; 128];
+    ///     let mut ancillary = SocketAncillary::new(&mut ancillary_buffer[..]);
+    ///
+    ///     let mut buf = [1; 8];
+    ///     let mut bufs = &mut [IoSliceMut::new(&mut buf[..])][..];
+    ///     sock.recv_vectored_with_ancillary(bufs, &mut ancillary)?;
+    ///
+    ///     ancillary.get_fds(&mut fds[..]);
+    ///     Ok(())
+    /// }
+    /// ```
+    #[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+    pub fn get_fds(&self, fds: &mut [RawFd]) -> usize {
```

Comment:
Rather than having all of these accessors to copy out various bits of information, I think it would be cleaner to offer an iterator interface over the individual messages.

Reply:
Can you show me how the Iterator should look like?

Reply:
Something like

```rust
struct CommandMessage([u8]);

impl CommandMessage {
    pub fn as_fd(&self) -> Option<RawFd> { ... }

    // etc for other accessors
}

pub struct Messages<'a> { ... }

impl<'a> Iterator for Messages<'a> {
    type Item = &'a CommandMessage;

    fn next(&mut self) -> Option<&'a CommandMessage> {
        // call CMSG_NEXTHDR etc
}

impl SocketAncillary {
    pub fn messages(&self) -> Message<'_> { .. }
}
```

Reply:
A cmsg can contain an array of FD and not a single FD.

Reply:
Sure, the message interface would be a bit different then.

Reply:
Okay, but what should be returned (`Option<Vec<RawFd>>`)?

Reply:
I'd probably actually name the accessor after the type:

```rust
impl CommandMessage {
    pub fn as_scm_rights(&self) -> Option<ScmRights<'_>> { ... }
}

pub struct ScmRights<'a>(&'a [u8]);

impl ScmRights {
    pub fn iter(&self) -> ScmRightsIter { ... }
}

pub struct ScmRightsIter(...);

impl Iterator for ScmRightsIter {
    type Item = RawFd;
}
```

Reply:
But then the `ScmRightsIter` would have to use `memcpy` for each `next` call, because the data is not aligned. This would be grossly inefficient.

Reply:
It's reading 4 bytes in each call to next. How many millions of file descriptors per second are you planning on transferring? How have you measured the penalty from making those loads unaligned, and how does that penalty compare to the cost of allocating a new, distinct memory region and copying the entire set of file descriptors to that?

Reply:
Okay, I would change that.

Reply:
Rather than a struct with a bunch of `as_foo(&self)` accessors, I still think it would be much better for the iterator to yield an enum as I proposed above, e.g.

```rust
#[non_exhaustive]
enum AncillaryData<'a> {
    ScmRights(ScmRights<'a>)
}
```

Reply:
Interesting. How would the iterator for the FDs would look like?

Reply:
It could look something like:
```rust
struct ScmRights<'a> {
    data: &'a [u8]
}
```
The `Iterator::next` implementation could be as simple as `Some(RawFd::from_ne_bytes([u8; mem::size_of::<RawFd>()]::try_from(&self.data[..mem::size_of::<RawFd>()]).ok()?))`.

Reply:
Okay, then the `ScmRights` struct would implement the `Iterator` trait or should I implement a dedicated Iterator like `ScmRightsIter`?

Reply:
I think making it an `Iterator` directly makes sense, but I don't have a strong opinion.

Reply:
@Ralith and @sfackler: I think there are too many structs defined here. To get a single `RawFd` you have to: 
`SocketAncillary` -> `Messages` -> `CommandMessage` -> `AncillaryData` -> `ScmRights` -> `RawFd`

I think `CommandMessage` is not necessary.

Reply:
My `enum AncillaryData` was intended as a replacement for @sfackler's `CommandMessage`.